src/Xamarin.Android.Tools.Bytecode/Kotlin/KotlinFixups.cs

```diff
@@ -83,15 +83,33 @@ static void FixupClassVisibility (ClassFile klass, KotlinClass metadata)
 				Log.Debug ($"Kotlin: Hiding internal class {klass.ThisClass.Name.Value}");
 				klass.AccessFlags = SetVisibility (klass.AccessFlags, ClassAccessFlags.Private);
 
-				foreach (var ic in klass.InnerClasses) {
-					Log.Debug ($"Kotlin: Hiding nested internal type {ic.InnerClass.Name.Value}");
-					ic.InnerClassAccessFlags = SetVisibility (ic.InnerClassAccessFlags, ClassAccessFlags.Private);
-				}
+				foreach (var ic in klass.InnerClasses)
+					HideInternalInnerClass (ic, classes);
 
 				return;
 			}
 		}
 
+		static void HideInternalInnerClass (InnerClassInfo klass, IList<ClassFile> classes)
+		{
+			var existing = klass.InnerClassAccessFlags;
+
+			klass.InnerClassAccessFlags = SetVisibility (existing, ClassAccessFlags.Private);
+			Log.Debug ($"Kotlin: Hiding nested internal type {klass.InnerClass.Name.Value}");
+
+			// Setting the inner class access flags above doesn't technically do anything, because we output
+			// from the inner class's ClassFile, so we need to find that and set it there.
+			var kf = classes.FirstOrDefault (c => c.ThisClass.Name.Value == klass.InnerClass.Name.Value);
+
+			if (kf != null) {
+				kf.AccessFlags = SetVisibility (kf.AccessFlags, ClassAccessFlags.Private);
+				kf.InnerClass.InnerClassAccessFlags = SetVisibility (kf.InnerClass.InnerClassAccessFlags, ClassAccessFlags.Private);
+
+				foreach (var inner_class in kf.InnerClasses.Where (c => c.OuterClass.Name.Value == kf.ThisClass.Name.Value))
```

Comment:
Shouldn't this `.Where()` cause *all* of `kf.InnerClasses` to be included?  How could `kf.InnerClasses` contain a type for which the inner type's `c.OuterClass` doesn't match the containing type?

Wouldn't this more easily & efficiently be expressed as:

```csharp
foreach (var inner_class in kf.InnerClasses) {
}
```

Reply:
…or does `kf.InnerClasses` contain "grand-children" as well?  For example, in the case of `HidePublicInterfaceInInternalClass`, will `InternalClassWithNestedInterface.class` specify `InternalClassWithNestedInterface$NestedInterface$DoubleNestedInterface.class` as an inner class, even though it's an inner class of `InternalClassWithNestedInterface$NestedInterface.class`?

Reply:
For some reason the `InnerClasses` collection for an inner class contains itself.  Without this check it infinitely recurses into itself.

That is, the `InnerClasses` for `InternalClassWithNestedInterface$NestedInterface` contains `InternalClassWithNestedInterface$NestedInterface`.